src/peft/tuners/vera/model.py

```diff
@@ -0,0 +1,564 @@
+
+        r = vera_config.r
+        bias = hasattr(target, "bias") and target.bias is not None
+        kwargs = {
+            "r": r,
+            "vera_dropout": vera_config.vera_dropout,
+            "fan_in_fan_out": vera_config.fan_in_fan_out,
+            "init_vera_weights": vera_config.init_vera_weights,
+        }
+
+        # TODO: add back once we have quant support
+        # kwargs["loaded_in_8bit"] = False
+        # kwargs["loaded_in_4bit"] = False
+        # quantization_config = get_quantization_config(self.model, method="gptq")
+        # if quantization_config is not None:
+        # kwargs["gptq_quantization_config"] = quantization_config
+
+        kwargs["bias"] = bias
+
+        # the below todo is copied from LoRA
+        # TODO: better deal with that
+        if isinstance(target, Embedding):
+            target.update_layer_embedding(
+                adapter_name,
+                r,
+                vera_config.vera_dropout,
+                vera_config.init_vera_weights,
+                d_initial=vera_config.d_initial,
+            )
+        elif isinstance(target, Linear):
+            target.update_layer(
+                adapter_name,
+                r,
+                vera_config.vera_dropout,
+                vera_config.init_vera_weights,
+                d_initial=vera_config.d_initial,
+            )
+        else:
+            new_module = self._create_new_module(vera_config, adapter_name, target, **kwargs)
+            if adapter_name != self.active_adapter:
+                # adding an additional adapter: it is not automatically trainable
+                new_module.requires_grad_(False)
+            self._replace_module(parent, target_name, new_module, target)
+
+    @staticmethod
+    def _replace_module(parent, child_name, new_module, child):
+        setattr(parent, child_name, new_module)
+        # It's not necessary to set requires_grad here, as that is handled by
+        # _mark_only_adapters_as_trainable
+
+        # child layer wraps the original module, unpack it
+        if hasattr(child, "base_layer"):
+            child = child.base_layer
+
+        # the below todo is copied from LoRA
+        # TODO: layers with base_layer don't need the weight to be copied, as they have a reference already
+        if not hasattr(new_module, "base_layer"):
+            new_module.weight = child.weight
+            if hasattr(child, "bias"):
+                new_module.bias = child.bias
+
+        if getattr(child, "state", None) is not None:
+            if hasattr(new_module, "base_layer"):
+                new_module.base_layer.state = child.state
+            else:
+                new_module.state = child.state
+            new_module.to(child.weight.device)
+
+        # dispatch to correct device
+        for name, module in new_module.named_modules():
+            if "vera_" in name:
+                module.to(child.weight.device)
+            if "ranknum" in name:
```

Comment:
Checking `ranknum` shouldn't be necessary. It only exists on LoRA because AdaLoRA recycles some of its code.